bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryDirectDataWriterHelper.java

```diff
@@ -253,6 +259,13 @@ public long finalizeStream() throws IOException {
     if (this.protoRows.getSerializedRowsCount() != 0) {
       sendAppendRowsRequest();
     }
+    try {
+      ApiFutures.allAsList(appendRowsFutures).get();
+    } catch (InterruptedException | ExecutionException e) {
+      throw new BigQueryConnectorException("Could not retrieve AppendRowsResponse", e);
+    } finally {
+      appendRowsExecutor.shutdown();
+    }
```

Comment:
A consequence of checking the async results here in finalizeStream() is that all (unless throttled by the write client library) rows will get sent regardless of an earlier failure. This could cause greater stress on the server. For example, if a request resulted in a resource-exceeded quota error, the additional rows would be sent anyway. One way to mitigate this is to check the status of any outstanding responses that may have already completed, at the beginning of sendAppendRowsRequest(). If a failure is detected at that time, we can throw an exception and cause an abort right away.

Reply:
This change has been made in https://github.com/GoogleCloudDataproc/spark-bigquery-connector/pull/1057